src/dialects/abstract/data-types.ts

```diff
@@ -0,0 +1,1826 @@
+  validate(value: Member): true {
+    if (!this.values.includes(value)) {
+      throw new ValidationError(
+        util.format('%j is not a valid choice in %j', value, this.values),
+        [],
+      );
+    }
+
+    return true;
+  }
+}
+
+/**
+ * An enumeration, Postgres Only
+ *
+ * @example
+ * DataTypes.ENUM('value', 'another value')
+ * DataTypes.ENUM(['value', 'another value'])
+ * DataTypes.ENUM({
+ *   values: ['value', 'another value']
+ * })
+ */
+export type ENUM<Member extends string> = _ENUM<Member>;
+export const ENUM = classToInvokable(_ENUM);
+
+interface ArrayOptions<T extends ABSTRACT<any>> {
+  type: T;
+}
+
+/**
+ * An array of `type`. Only available in Postgres.
+ *
+ * @example
+ * DataTypes.ARRAY(DataTypes.DECIMAL)
+ */
+class _ARRAY<T extends ABSTRACT<any>> extends ABSTRACT<
+  Array<AcceptableTypeOf<T>>,
+  Array<SaneTypeOf<T>>,
+  Array<RawTypeOf<T>>
+> {
+  public static readonly key = 'ARRAY';
+  protected options: ArrayOptions<T>;
+  public type: T;
+
+  /**
+   * @param type type of array values
+   */
+  constructor(type: DataType<T> | { type: DataType<T> }) {
+    super();
+    const opts = ABSTRACT.isType(type) ? { type } : type;
+
+    this.options = {
+      type: typeof opts.type === 'function' ? new opts.type() : opts.type,
+    };
+
+    this.type = this.options.type;
+  }
+
+  toSql(options: StringifyOptions) {
+    return `${this.type.toSql(options)}[]`;
+  }
+
+  // Note: Validation of individual items is handled in the query-generator.
+  public validate(value: Array<AcceptableTypeOf<T>>) {
+    if (!Array.isArray(value)) {
+      throw new ValidationError(
+        util.format('%j is not a valid array', value),
+        [],
+      );
+    }
+
+    return true;
+  }
+
+  static is<T extends ABSTRACT<any>>(
+    obj: unknown,
+    type: new () => T,
+  ): obj is ARRAY<T> {
+    return obj instanceof ARRAY && (obj as ARRAY<any>).type instanceof type;
+  }
+}
+
+export type ARRAY<T extends ABSTRACT<any>> = _ARRAY<T>;
+export const ARRAY = classToInvokable(_ARRAY);
+
+export type GeometryType = Uppercase<keyof typeof wkx>;
+
+interface GeometryOptions<Type extends GeometryType> {
+  type?: Type;
+  srid?: number;
+}
+
+export type GeometryParams<Type extends GeometryType> =
+  | []
+  | [GeometryOptions<Type> | undefined]
+  | [Type | undefined]
+  | [Type | undefined, number | undefined];
+
+class _GEOMETRY<Type extends GeometryType = GeometryType> extends ABSTRACT<
+  wkx.Geometry | Buffer | string,
+  wkx.Geometry
+> {
+  public static readonly key: string = 'GEOMETRY';
+  public static readonly escape = false;
+  protected options: GeometryOptions<Type>;
+  protected type?: Type;
+  protected srid?: number;
+
+  /**
+   * @param {string} [type] Type of geometry data
+   * @param {string} [srid] SRID of type
+   */
+  constructor(type: Type, srid?: number);
+  constructor(options: GeometryOptions<Type>);
+  constructor(...args: GeometryParams<Type>);
```

Comment:
using a rest parameter here, when you don't actually accept rest params, is misleading. Plus this seems to be already covered by the previous 2 overloads